src/site/content/en/blog/subsonic/index.md

```diff
@@ -0,0 +1,171 @@
+// state is an object containing two fields: the array buffer in 'buffer' and
+// the current input index in 'index'.
+function readTarHeader(state) {
+// the offset of the file this header describes is always 512 bytes from
+// the start of the header
+var offset = state.index + 512;
+
+// the header is made up of several fields at fixed offsets within the
+// 512 byte block allocated for the header.  fields have a fixed length.
+// all numeric fields are stored as octal numbers encoded as ASCII
+// strings.
+var name = readString(state, 100);
+var mode = parseInt(readString(state, 8), 8);
+var uid = parseInt(readString(state, 8), 8);
+var gid = parseInt(readString(state, 8), 8);
+var size = parseInt(readString(state, 12), 8);
+var modified = parseInt(readString(state, 12), 8);
+var crc = parseInt(readString(state, 8), 8);
+var type = parseInt(readString(state, 1), 8);
+var link = readString(state, 100);
+
+// the header is followed by the file contents, then followed
+// by padding to ensure that the next header is on a 512-byte
+// boundary.  advanced the input state index to the next
+// header.
+state.index = offset + Math.ceil(size / 512) * 512;
+
+// return the descriptor with the relevant fields we care about
+return {
+name : name,
+size : size,
+type : type,
+offset : offset
+};
+};
+```
+
+I looked around for existing TAR readers, and found a few, but none that did not have other dependencies or which would easily fit into our existing codebase.  For this reason, I chose to write my own.  I also took the time to optimize the loading as best as possible, and ensure that the decoder easily handles both binary and string data within the archive.
+
+One of the first problems I had to solve was how to actually get the data loaded from an XHR request.  I originally started with a "binary string" approach, which make parsing the TAR headers quite easily.  Unfortunately, converting from binary strings to more readily usable binary forms like an `ArrayBuffer` is not straight forward, nor are such conversions particularly quick.  Converting to `Image` objects is equally painful.
+
+I settled on loading the TAR files as an `ArrayBuffer` directly from the XHR request and adding a small convenience function for converting chunks from the `ArrayBuffer` to a string.  Currently my code only handles basic ANSI/8-bit characters, but this can be fixed once a more convenient conversion API is available in browsers.
+
+The code simply scans over the `ArrayBuffer` parsing out record headers, which includes all of the relevant TAR header fields (and a few not-so-relevant ones) as well as the location and size of the file data within the `ArrayBuffer`.  The code can also optionally extract the data as an `ArrayBuffer` view and store that in the returned record headers list.
+
+The code is freely available under a friendly, permissive Open Source license at [https://github.com/subsonicllc/TarReader.js](https://github.com/subsonicllc/TarReader.js).
```

Comment:
```suggestion
I looked around for existing TAR readers, and found a few, but none that did not have other dependencies or which would easily fit into our existing codebase. For this reason, I chose to write my own. I also took the time to optimize the loading as best as possible, and ensure that the decoder easily handles both binary and string data within the archive.

One of the first problems I had to solve was how to actually get the data loaded from an XHR request. I originally started with a "binary string" approach. Unfortunately, converting from binary strings to more readily usable binary forms like an `ArrayBuffer` is not straightforward, nor are such conversions particularly quick. Converting to `Image` objects is equally painful.

I settled on loading the TAR files as an `ArrayBuffer` directly from the XHR request and adding a small convenience function for converting chunks from the `ArrayBuffer` to a string. Currently my code only handles basic ANSI/8-bit characters, but this can be fixed once a more convenient conversion API is available in browsers.

The code simply scans over the `ArrayBuffer` parsing out record headers, which includes all of the relevant TAR header fields (and a few not-so-relevant ones) as well as the location and size of the file data within the `ArrayBuffer`. The code can also optionally extract the data as an `ArrayBuffer` view and store that in the returned record headers list.

The code is freely available under a friendly, permissive Open Source license at [https://github.com/subsonicllc/TarReader.js](https://github.com/subsonicllc/TarReader.js).
```